// @temporalio/nexus/lib/workflow-helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemporalOperationHandler = exports.TemporalOperationResult = exports.WorkflowRunOperationHandler = void 0;
exports.startWorkflow = startWorkflow;
const nexus = __importStar(require("nexus-rpc"));
const internal_1 = require("@temporalio/client/lib/internal");
const link_converter_1 = require("./link-converter");
const token_1 = require("./token");
const context_1 = require("./context");
/**
 * Starts a workflow run for a {@link WorkflowRunOperationStartHandler}, linking the execution chain
 * to a Nexus Operation (subsequent runs started from continue-as-new and retries). Automatically
 * propagates the callback, request ID, and back and forward links from the Nexus options to the
 * Workflow.
 *
 * @experimental Nexus support in Temporal SDK is experimental.
 */
async function startWorkflow(ctx, workflowTypeOrFunc, workflowOptions) {
    const { client, taskQueue } = (0, context_1.getHandlerContext)();
    const links = Array();
    if (ctx.inboundLinks?.length > 0) {
        for (const l of ctx.inboundLinks) {
            try {
                links.push((0, link_converter_1.convertNexusLinkToTemporalLink)(l));
            }
            catch (error) {
                context_1.log.warn('failed to convert Nexus link to Workflow event link', { error });
            }
        }
    }
    const internalOptions = {
        links,
        requestId: ctx.requestId,
    };
    internalOptions.onConflictOptions = {
        attachLinks: true,
        attachCompletionCallbacks: true,
        attachRequestId: true,
    };
    // Add nexus-operation-token header to solve for race between Workflow completion
    // and Nexus Operation start recording
    const callbackHeaders = {
        ...ctx.callbackHeaders,
        'nexus-operation-token': (0, token_1.generateWorkflowRunOperationToken)(client.options.namespace, workflowOptions.workflowId),
    };
    if (ctx.callbackUrl) {
        internalOptions.completionCallbacks = [
            {
                nexus: { url: ctx.callbackUrl, header: callbackHeaders },
                links, // pass in links here as well for older servers, newer servers dedupe them.
            },
        ];
    }
    const { taskQueue: userSpecifiedTaskQueue, ...rest } = workflowOptions;
    const startOptions = {
        ...rest,
        taskQueue: userSpecifiedTaskQueue || taskQueue,
        [internal_1.InternalWorkflowStartOptionsSymbol]: internalOptions,
    };
    const handle = await client.workflow.start(workflowTypeOrFunc, startOptions);
    if (internalOptions.backLink != null) {
        try {
            ctx.outboundLinks.push((0, link_converter_1.convertTemporalLinkToNexusLink)(internalOptions.backLink));
        }
        catch (error) {
            context_1.log.warn('failed to convert temporal link to Nexus link', { error });
        }
    }
    return {
        workflowId: handle.workflowId,
        runId: handle.firstExecutionRunId,
    };
}
/**
 * A Nexus Operation implementation that is backed by a Workflow run.
 *
 * @experimental Nexus support in Temporal SDK is experimental.
 */
class WorkflowRunOperationHandler {
    handler;
    constructor(handler) {
        this.handler = handler;
    }
    async start(ctx, input) {
        const { namespace } = (0, context_1.getHandlerContext)();
        const handle = await this.handler(ctx, input);
        return nexus.HandlerStartOperationResult.async((0, token_1.generateWorkflowRunOperationToken)(namespace, handle.workflowId));
    }
    async cancel(_ctx, token) {
        const decoded = (0, token_1.loadWorkflowRunOperationToken)(token);
        await (0, context_1.getClient)().workflow.getHandle(decoded.wid).cancel();
    }
}
exports.WorkflowRunOperationHandler = WorkflowRunOperationHandler;
/**
 * Module-private brand and payload key for {@link TemporalOperationResult}.
 */
const operationResult = Symbol('temporal_nexus_TemporalOperationResult');
exports.TemporalOperationResult = {
    sync(value) {
        return {
            [operationResult]: nexus.HandlerStartOperationResult.sync(value),
        };
    },
    async(token) {
        return {
            [operationResult]: nexus.HandlerStartOperationResult.async(token),
        };
    },
};
class TemporalNexusClientImpl {
    startOperationContext;
    asyncOperationStarted = false;
    constructor(startOperationContext) {
        this.startOperationContext = startOperationContext;
    }
    /**
     * The Temporal Client for the active Nexus Operation.
     *
     * @experimental Nexus support in Temporal SDK is experimental.
     */
    get client() {
        return (0, context_1.getClient)();
    }
    /**
     * Starts a workflow run as the asynchronous backing operation for the current Nexus Operation.
     *
     * @experimental Nexus support in Temporal SDK is experimental.
     */
    async startWorkflow(workflowTypeOrFunc, workflowOptions) {
        return await this.withAsyncOperationStartReservation(async () => {
            const handle = await startWorkflow(this.startOperationContext, workflowTypeOrFunc, workflowOptions);
            const { namespace } = (0, context_1.getHandlerContext)();
            return exports.TemporalOperationResult.async((0, token_1.generateWorkflowRunOperationToken)(namespace, handle.workflowId));
        });
    }
    async withAsyncOperationStartReservation(fn) {
        if (this.asyncOperationStarted) {
            throw new nexus.HandlerError('BAD_REQUEST', 'Only one async operation can be started per operation handler invocation. Use TemporalNexusClient.client for additional workflow interactions');
        }
        this.asyncOperationStarted = true;
        try {
            return await fn();
        }
        catch (err) {
            this.asyncOperationStarted = false;
            throw err;
        }
    }
}
/**
 * A Nexus Operation implementation for operations that interact with Temporal.
 *
 * @experimental Nexus support in Temporal SDK is experimental.
 */
class TemporalOperationHandler {
    startHandler;
    cancelWorkflowRunHandler;
    constructor(options) {
        this.startHandler = options.start;
        this.cancelWorkflowRunHandler = options.cancelWorkflowRun ?? defaultCancelWorkflowRun;
    }
    async start(ctx, input) {
        const result = await this.startHandler(ctx, new TemporalNexusClientImpl(ctx), input);
        return result[operationResult];
    }
    async cancel(ctx, token) {
        let opToken;
        try {
            opToken = (0, token_1.loadOperationToken)(token);
        }
        catch (err) {
            throw new nexus.HandlerError(nexus.HandlerErrorType.BAD_REQUEST, 'invalid operation token', { cause: err });
        }
        switch (opToken.t) {
            case token_1.OperationTokenType.WORKFLOW_RUN:
                try {
                    (0, token_1.assertWorkflowRunOperationToken)(opToken);
                }
                catch (err) {
                    throw new nexus.HandlerError(nexus.HandlerErrorType.BAD_REQUEST, 'invalid workflow run operation token', {
                        cause: err,
                    });
                }
                await this.cancelWorkflowRunHandler(ctx, { workflowId: opToken.wid });
                return;
            default:
                throw new nexus.HandlerError(nexus.HandlerErrorType.BAD_REQUEST, `Unsupported operation token type: ${opToken.t}`);
        }
    }
}
exports.TemporalOperationHandler = TemporalOperationHandler;
async function defaultCancelWorkflowRun(_ctx, options) {
    await (0, context_1.getClient)().workflow.getHandle(options.workflowId).cancel();
}
